include/Edge.hpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of the JKQ DD Package which is released under the MIT license.
+ * See file README.md or go to http://iic.jku.at/eda/research/quantum_dd/ for more information.
+ */
+
+#ifndef DD_PACKAGE_EDGE_HPP
+#define DD_PACKAGE_EDGE_HPP
+
+#include "Complex.hpp"
+#include "ComplexTable.hpp"
+#include "ComplexValue.hpp"
+
+#include <cstddef>
+#include <utility>
+
+namespace dd {
+    template<class Node>
+    struct Edge {
+        Node*   p;
+        Complex w;
+
+        constexpr bool operator==(const Edge& other) const {
+            return p == other.p && w.approximatelyEquals(other.w);
+        }
+        constexpr bool operator!=(const Edge& other) const {
+            return !operator==(other);
+        }
+
+        [[nodiscard]] constexpr bool isTerminal() const { return Node::isTerminal(p); }
+
+        // edges pointing to zero and one terminals
+        static inline Edge one{Node::terminal, Complex::one};
+        static inline Edge zero{Node::terminal, Complex::zero};
+
+        [[nodiscard]] static constexpr Edge terminal(const Complex& w) { return {Node::terminal, w}; }
+        [[nodiscard]] constexpr bool        isZeroTerminal() const { return Node::isTerminal(p) && w == Complex::zero; }
+        [[nodiscard]] constexpr bool        isOneTerminal() const { return Node::isTerminal(p) && w == Complex::one; }
+    };
+
+    template<typename Node>
+    struct CachedEdge {
+        Node*        p{};
+        ComplexValue w{};
+
+        CachedEdge() = default;
+        CachedEdge(Node* p, const ComplexValue& w):
+            p(p), w(w) {}
+        CachedEdge(Node* p, const Complex& c):
+            p(p) {
+            w.r = c.r->val();
+            w.i = c.i->val();
+        }
+        bool operator==(const CachedEdge& other) const {
+            return p == other.p && w.approximatelyEquals(other.w);
```

Comment:
I think its rather surprising that `==` checks the equivalence approximately. Maybe we can discuss this in the next meeting.

Reply:
I'll resolve this comment for now and leave the discussions in the one above.